### PR TITLE
HIVE-24503 : Optimize vector row serde by avoiding type check at run time.

### PR DESCRIPTION
**Pull request by maheshk114**
…

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java [hunk 1]
```diff
@@ -61,27 +61,16 @@
...
+    Field[] children = null;
+    boolean isPrimitive = false;
+    Category category = null;
+    PrimitiveCategory primitiveCategory = null;
+    TypeInfo typeInfo = null;
+    int count = 0;
+    ObjectInspector objectInspector = null;
+    int outputColumnNum = -1;
+    VectorSerializeWriter writer = null;
     Field() {
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:73]: Can be removed.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java [hunk 2]
```diff
@@ -933,12 +1207,20 @@ private void storeUnionRowColumn(ColumnVector colVector,
...
         unionHelper.getFields()[tag],
         batchIndex,
         canRetainByteRef);
-    deserializeRead.finishComplexVariableFieldsType();
+    deserializer.deserializeRead.finishComplexVariableFieldsType();
+  }
+
+  abstract static class VectorBatchDeserializer {
+    abstract void store(ColumnVector colVector, Field field, int batchIndex, boolean canRetainByteRef,
+                            VectorDeserializeRow deserializer) throws IOException;
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java:1220]: Why VectorDeserializerRow needs be passed here again? ("this" references in other places as well). If you remove "static" class declaration in VectorBatchDeserializer children, you may not need to pass this.  And the patch would become lot lesser changes?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java [hunk 3]
```diff
@@ -274,44 +315,25 @@ private void serializeWrite(
...
-      serializeUnionWrite(
-          (UnionColumnVector) colVector,
-          field,
-          adjustedBatchIndex);
-      break;
-    default:
-      throw new RuntimeException("Unexpected category " + category);
+  abstract static class VectorSerializeWriter {
+    abstract void serialize(Object colVector, Field field, int adjustedBatchIndex,
+                            VectorSerializeRow serializeRow) throws IOException;
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:323]: Same as earlier. VectorSerializeRow need not be passed here. Patch may need lesser changes if you remove static declaration on children.

**Review by rbalamohan — APPROVED** [review]
Thanks @maheshk114. LGTM. +1.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java [hunk 4]
```diff
@@ -328,6 +345,13 @@ private void serializeUnionWrite(
...
   }
 
+  class VectorSerializeStructWriter extends VectorSerializeWriter {
+    @Override
+    void serialize(Object colInfo, Field field, int adjustedBatchIndex) throws IOException {
+      serializeStructWrite((StructColumnVector)colInfo, field, adjustedBatchIndex);
+    }
+  }
+
   private void serializeStructWrite(
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:355]: Nit: would be cleaner if serializeStructWrite method is be part of the VectorSerializeStructWriter class

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java [hunk 5]
```diff
@@ -355,6 +379,13 @@ private void serializeStructWrite(
...
   }
 
+  class VectorSerializeMapWriter extends VectorSerializeWriter {
+    @Override
+    void serialize(Object colInfo, Field field, int adjustedBatchIndex) throws IOException {
+      serializeMapWrite((MapColumnVector)colInfo, field, adjustedBatchIndex);
+    }
+  }
+
   private void serializeMapWrite(
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:389]: Nit: serializeMapWrite part of VectorSerializeMapWriter class?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java [hunk 6]
```diff
@@ -383,6 +414,13 @@ private void serializeMapWrite(
...
   }
 
+  class VectorSerializeListWriter extends VectorSerializeWriter {
+    @Override
+    void serialize(Object colInfo, Field field, int adjustedBatchIndex) throws IOException {
+      serializeListWrite((ListColumnVector)colInfo, field, adjustedBatchIndex);
+    }
+  }
+
   private void serializeListWrite(
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:424]: Nit: serializeListWrite part of VectorSerializeListWriter?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java [hunk 7]
```diff
@@ -564,173 +621,337 @@ public void init() throws HiveException {
...
+      storeString(colVector, field, batchIndex, canRetainByteRef);
+    }
+
+    @Override
+    Object convert(ColumnVector batch, int batchIndex, Field field) throws IOException {
+      return convertString(field.getConversionWritable(), batchIndex);
+    }
+  }
+
+  class VectorVarcharDeserializer extends VectorBatchDeserializer {
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java:808]: Looks like VectorVarcharDeserializer and VectorCharDeserializer share the same store method -- so maybe make them inherit from a common base class and only Override convert method?

**Review by pgaref — CHANGES_REQUESTED** [review]
Impressed by how cleaner this looks :)
Left some comments -- @maheshk114  let me know what you think

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java (hunk 4 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:355]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java (hunk 5 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:389]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java (hunk 6 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorSerializeRow.java:424]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java (hunk 7 above)
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorDeserializeRow.java:808]: i think the code duplication is not much and the classes are too simple. creating a common class will be over engineering.

**Review by pgaref — APPROVED** [review]
Thanks for the changes @maheshk114 !

**Pull request merged**